source/OptimaJet.DWKit.StarterApplication/Controllers/BaseController.cs

```diff
@@ -39,14 +41,33 @@ public User CurrentUser
                 // current user has not yet been found for this request.
                 // find or create because users are managed by auth0 and
                 // creation isn't proxied through the api.
-                var auth0Id = HttpContext.GetAuth0Id();
-                var user = userService.FindOrCreateUser(auth0Id).Result;
+                var user = FindOrCreateCurrentUser();
 
                 HttpContext.Items[CURRENT_USER_KEY] = user;
 
                 return user;      
             }
             
         }
+
+        private User FindOrCreateCurrentUser()
```

Comment:
I'd move this directly into the CurrentUser Context since it's more of a concern of there than the _base_ controller.

```cs
private User currentUser = null; // or _currentUser, whatever we decided on earlier
public User GetCurrentUser() 
{
     if (!currentUser)
     {
         currentUser = CurrentUserContext.FindOrCreateCurrentUser();
     }

    return currentUser;
}
```

and then in the base controller, you could just do:

```cs
protected User CurrentUser 
{
  get 
  {
    return CurrentUserContext.GetCurrentUser();
  }
}
```

This also would allow the CurrentUserContext to be injected anywhere and allow any class to have memoized access to the current user.

Reply:
HttpCurrentUserContext only read the information from the HttpContext (and Auth0 Management API) and doesn't have any dependencies on Model objects or other services.  It would be an inappropriate coupling.

Reply:
CurrentUserContext could directly interact with the AppDbContext instead of UserService, I think that would solve the problem :)

I'd do:

CurrentUserContext - depends on:
 - IHttpContextAccessor
 - Auth0Management
 - AppDbContext

UserService (and any other Service) could then inject the CurrentUserContext without circular injection.

At least, that's what I think would provide the best ergonomics, given all the research / messin' with things so far.

Thoughts?

Reply:
I still like HttpCurrentUserContext being low-level (just above the Network (HttpContext/Auth0 Management API)).  There doesn't seem to be value in pushing it up to contain model objects.

To me the best place to FindOrCreateUser is at the BaseController.